【PaddlePaddle Hackathon 2】8、为 Paddle 新增 nanmean API

Li-fAngyU opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types 
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
New features
### PR changes
<!-- One of [ OPs | APIs | Docs | Others ] -->
APIs
### Describe
<!-- Describe what this PR does -->
ISSUE链接:https://github.com/PaddlePaddle/Paddle/issues/40327
RFC的PR链接:https://github.com/PaddlePaddle/community/pull/48
中文文档PR链接:https://github.com/PaddlePaddle/docs/pull/4294
为 Paddle 新增 nanmean API

paddle.nanmean 扩展了 paddle.mean API 的功能，如果输入Tensor中有nan值， paddle.mean在计算中会将涉及nan值的结果都置为nan，而 paddle.nanmean 会跳过nan值。比如输入数据 x = [[nan, 1. , 2. ], [3. , 4. , 5. ]]，x.mean() 得到 [nan]，x.mean(0) 得到 [nan, 2.5, 3.5]，x.nanmean() 得到 [3.]，x.nanmean(0) 得到 [3., 2.5, 3.5]。此API需支持的调用路径为：paddle.nanmean 和 Tensor.nanmean 。

Li-fAngyU committed:
Update __init__.py

Li-fAngyU committed:
Update math.py

Li-fAngyU committed:
Create test_nanmean_api.py

Li-fAngyU committed:
Update __init__.py

Li-fAngyU committed:
Update __init__.py

Li-fAngyU committed:
Update math.py

Li-fAngyU committed:
Update test_nanmean_api.py

Li-fAngyU committed:
Update __init__.py

Li-fAngyU committed:
Update math.py

Li-fAngyU committed:
Update test_nanmean_api.py

Li-fAngyU committed:
Update test_nanmean_api.py

Li-fAngyU committed:
Update test_nanmean_api.py

Li-fAngyU committed:
Update math.py

Li-fAngyU committed:
Update test_nanmean_api.py

Li-fAngyU committed:
Update math.py

Update the nanmean example code

Li-fAngyU committed:
Update math.py

Li-fAngyU committed:
Update math.py

jeff41404 commented on python/paddle/tensor/math.py:
```diff
@@ -967,6 +967,81 @@ def nansum(x, axis=None, dtype=None, keepdim=False, name=None):
     return sum(tmp_tensor, axis, dtype, keepdim, name)
 
 
+def nanmean(x,axis=None,keepdim=None,name=None):
```
keepdim=False is better？keep same with other API？

Li-fAngyU replied:
thank you for your suggestion! keep same with other API is the better way. i will update this

jeff41404 commented on python/paddle/tensor/math.py:
```diff
@@ -967,6 +967,81 @@ def nansum(x, axis=None, dtype=None, keepdim=False, name=None):
+        axis = [axis]
+    if axis == None:
+        return paddle.mean(x[~paddle.isnan(x)], keepdim=keepdim,name=name)
+    check_variable_and_dtype(x, 'x/input',
```
just 'x' instead of  'x/input'?

Li-fAngyU replied:
this is refer to the code of [paddle.mean](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/tensor/stat.py#L94).
'x/input' seems to be used only as a input name when raise a type/dtype error.

jeff41404 commented on python/paddle/tensor/math.py:
```diff
@@ -967,6 +967,81 @@ def nansum(x, axis=None, dtype=None, keepdim=False, name=None):
+    if axis == None:
+        return paddle.mean(x[~paddle.isnan(x)], keepdim=keepdim,name=name)
+    check_variable_and_dtype(x, 'x/input',
+                             ['uint16', 'float16', 'float32', 'float64'],
```
dtype should be the same to description of x above. eg: 'uint16' is not in "x (Tensor): The input Tensor with data type float32, float64."

Li-fAngyU replied:
This is refer to the code of [paddle.mean](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/tensor/stat.py#L35)
Although it require the input tensor must with data type of  float32/64, but it also check allow data type of 'uint16' and 'float16' [code is here](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/tensor/stat.py#L95).
Because of the issus describe [here](https://github.com/PaddlePaddle/Paddle/issues/40327)(paddle.nanmean extends the functionality of the paddle.mean API), so i just follow the code of paddle.mean

jeff41404 commented on python/paddle/tensor/math.py:
```diff
@@ -967,6 +967,81 @@ def nansum(x, axis=None, dtype=None, keepdim=False, name=None):
+    """
+    if isinstance(axis, int):
+        axis = [axis]
+    if axis == None:
+        return paddle.mean(x[~paddle.isnan(x)], keepdim=keepdim,name=name)
```
can the logic of code L1041~L1042 below cover this branch? we need to handle the condition 'axis == None' alone?

Li-fAngyU replied:
thank you for your suggestion! 
the logic of code L1041~L1042 below can cover this branch.
we don't need handle this condition alone.
i will take this advise and update this problem.

jeff41404 commented on python/paddle/fluid/tests/unittests/test_nanmean_api.py:
```diff
@@ -0,0 +1,87 @@
+
+    def setUp(self):
+        self.x_shape = [2, 3, 4, 5]
+        self.x = np.random.uniform(-1, 1, self.x_shape).astype(np.float32)
```
self.x does not have 'nan', we should cover all the test case in [rfcs](https://github.com/PaddlePaddle/community/pull/23/files)
also should include check gradient

Li-fAngyU replied:
thanks for your suggestion! i will cover all the test case in next update. 
but i am confused in check gradient, because i can't find the example of check gradient in the paddle.mean test file [test_mean_op.py](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/tests/unittests/test_mean_op.py) .
I will appreciate it if you can give me a example.

Ligoml commented:
hi~有些细节需要注意一下 @Li-fAngyU 
1、PR的标题要和ISSUE标题内容一样；
2、Describe 里要加上 ISSUE 链接 & RFC链接 & 中文文档链接

Li-fAngyU commented:
@Ligoml 谢谢提醒，我已按照您的提醒进行了修改。但是我的中文文档PR还未通过可以先附上PR链接吗？

Ligoml commented:
要关联起来的~这个有没有merge不影响的

Li-fAngyU commented:
好的

Li-fAngyU committed:
Update math.py

Remove redundant code in nanmean

Li-fAngyU committed:
Update math.py

change default keepdim = False

Li-fAngyU committed:
Update test_nanmean_api.py

add nan into self.x

Li-fAngyU committed:
Update test_nanmean_api.py

rerun CI check

Li-fAngyU committed:
Update test_nanmean_api.py

Li-fAngyU committed:
update code of nanmean in python/paddle/tensor/math.py and test_nanmean_api.py

Li-fAngyU committed:
Merge branch 'Li-fAngyU-patch-1' of https://github.com/Li-fAngyU/Paddle into Li-fAngyU-patch-1

Li-fAngyU committed:
Merge branch 'PaddlePaddle:develop' into Li-fAngyU-patch-1

Li-fAngyU committed:
Update test_nanmean_api.py

update code format

Li-fAngyU reviewed (commented):
@jeff41404  thank you for your reviewing and suggesion!!!
1. Update some code in math.py, but still have some problem confuse me? Should I check the variables and their dtypes as test_mean_op.py does.
2. add test nan value cases in test_nanmean_api.py, but i don't know how to check gradient? i will appreciate if you can give me a example.

Li-fAngyU committed:
Merge branch 'PaddlePaddle:develop' into Li-fAngyU-patch-1

Li-fAngyU committed:
Update test_nanmean_api.py

update code format

jeff41404 commented:
> @jeff41404 thank you for your reviewing and suggesion!!!
> 
> 1. Update some code in math.py, but still have some problem confuse me? Should I check the variables and their dtypes as test_mean_op.py does.
> 2. add test nan value cases in test_nanmean_api.py, but i don't know how to check gradient? i will appreciate if you can give me a example.
1. it's better that checklist and description are the same, so the description of paddle.mean need to be update, if have time, we encourage you update
2. because this API is composed of other APIs, we can use some fix data to check grad

Li-fAngyU commented:
i got it! thank you for your advise!

Li-fAngyU commented:
@jeff41404 Hi，when i check grad of nanmean,  i find a new issue of paddle.sum() in gradient. And because my API implementation uses this function, I also get this error when I run gradient tests. 
I would appreciate it if you could take a look at the issue [here](https://github.com/PaddlePaddle/Paddle/issues/40850).

Li-fAngyU committed:
Merge branch 'PaddlePaddle:develop' into Li-fAngyU-patch-1

Li-fAngyU committed:
Update test_nanmean_api.py

add check grad code.

Li-fAngyU committed:
Update math.py

update nanmean's describe of Args x

Li-fAngyU committed:
Merge branch 'PaddlePaddle:develop' into Li-fAngyU-patch-1

Li-fAngyU committed:
Update test_nanmean_api.py

update format and release the test_case(self.x, keepdim=True) in check grad code.

Li-fAngyU commented:
1. update the describe of nanmean's parameter.
2. Add code to check gradient in function test_api_dygraph_grad

jeff41404 commented on python/paddle/fluid/tests/unittests/test_nanmean_api.py:
```diff
@@ -0,0 +1,124 @@
+            sum_dx_ref = np.prod(y.shape)
+            if np.isnan(y.numpy()).sum():
+                sum_dx_ref -= np.isnan(y.numpy()).sum()
+            dx[np.isnan(dx)] = 0
```
dx[np.isnan(dx)] must be 0, unless all element in axis is nan. so set to 0 may cover the error or is not appropriate

jeff41404 commented on python/paddle/fluid/tests/unittests/test_nanmean_api.py:
```diff
@@ -0,0 +1,124 @@
+                sum_dx_ref -= np.isnan(y.numpy()).sum()
+            dx[np.isnan(dx)] = 0
+            sum_dx = dx.sum()
+            self.assertEqual(np.allclose(sum_dx, sum_dx_ref, rtol=1e-04), True)
```
this method to check grad may work in this data : self.x[0, :, :, :] = np.nan，but may not work in another
e.g. we have  x = [[float('nan'), 2. , 3. ], [0. , float('nan') , 2. ]], use p_x = paddle.to_tensor(x) and paddle.nanmean(p_x, 0) will be [0.0000, 2.0000, 2.5000], and after backward, x.grad should be  [[0.0000, 1.0000, 0.5000], [1.0000, 0.0000, 0.5000]]
another case: we have  x = [[float('nan'), float('nan') , 3. ], [0. , float('nan') , 2. ]], use p_x = paddle.to_tensor(x) and paddle.nanmean(p_x, 0) will be [0.0000, nan, 2.5000], and after backward, x.grad should be  [[0.0000, nan, 0.5000], [1.0000, nan, 0.5000]]

Li-fAngyU committed:
Merge branch 'PaddlePaddle:develop' into Li-fAngyU-patch-1

Li-fAngyU committed:
Update test_nanmean_api.py

Update gradient checking method

Li-fAngyU commented:
Done！thank you for your suggestion！

Li-fAngyU committed:
Merge branch 'PaddlePaddle:develop' into Li-fAngyU-patch-1

Li-fAngyU committed:
Merge branch 'PaddlePaddle:develop' into Li-fAngyU-patch-1

Li-fAngyU committed:
Update test_nanmean_api.py

update code format

jeff41404 reviewed (dismissed):
LGTM

Ligoml commented on python/paddle/tensor/math.py:
```diff
@@ -981,6 +981,80 @@ def nansum(x, axis=None, dtype=None, keepdim=False, name=None):
+        Tensor, results of arithmetic mean along ``axis`` of ``x``, with the same data
+        type as ``x``.
+
+    Examples:
```
需要与code空一行，保证文档样式正确显示。

Ligoml commented on python/paddle/fluid/tests/unittests/test_nanmean_api.py:
```diff
@@ -0,0 +1,136 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

Ligoml commented on python/paddle/tensor/math.py:
```diff
@@ -981,6 +981,80 @@ def nansum(x, axis=None, dtype=None, keepdim=False, name=None):
+    Compute the arithmetic mean along the specified axis, ignoring NaNs.
+
+    Args:
+        x (Tensor): The input Tensor with data type uint16, float16, float32, float64.
```
data type uint16, float16, float32, float64.
和中文文档不一致，需要确认数据类型

Li-fAngyU replied:
数据类型应该是 unit16, float16, float32, float64。我会修改中文文档和英文文档对齐。

Ligoml commented on python/paddle/tensor/math.py:
```diff
@@ -981,6 +981,80 @@ def nansum(x, axis=None, dtype=None, keepdim=False, name=None):
+            should be in range [-D, D), where D is the dimensions of ``x`` . If
+            ``axis`` or element(s) of ``axis`` is less than 0, it works the
+            same way as :math:`axis + D` . If ``axis`` is None, mean is
+            calculated over all elements of ``x``. Default is None.
```
建议改成：
If axis is None, nanmean is calculated over all elements of x. 
nanmean做主语

Ligoml replied:
@Li-fAngyU 这句中出现了两次 nanmean is calculated，需要同步修改哈

Ligoml commented on python/paddle/tensor/math.py:
```diff
@@ -981,6 +981,80 @@ def nansum(x, axis=None, dtype=None, keepdim=False, name=None):
+        .. code-block:: python
+
+            import paddle
+            import numpy as np
```
直接使用paddle.to_tensor生成输入数据，不需要import numpy

Li-fAngyU committed:
Merge branch 'PaddlePaddle:develop' into Li-fAngyU-patch-1

Li-fAngyU committed:
Update test_nanmean_api.py 

Update code format and copyright in test_nanmean_api.py

Li-fAngyU committed:
Update math.py

update arguments describe and code example

Li-fAngyU commented:
@Ligoml Done!

Li-fAngyU committed:
Update math.py

修改了nanmean的axis参数的文档描述。

Ligoml reviewed (dismissed):
LGTM for docs

Li-fAngyU committed:
Merge branch 'PaddlePaddle:develop' into Li-fAngyU-patch-1

Li-fAngyU committed:
Update math.py

updata nanmean's sample code (:name: code-example1)

XiaoguangHu01 reviewed (dismissed):
LGTM

Li-fAngyU committed:
Merge branch 'PaddlePaddle:develop' into Li-fAngyU-patch-1

Li-fAngyU committed:
Update math.py

修改nanmean的example code 错误

Ligoml reviewed (approved):
LGTM for docs

Li-fAngyU committed:
Merge branch 'PaddlePaddle:develop' into Li-fAngyU-patch-1

Li-fAngyU committed:
Update math.py

update example code

Li-fAngyU committed:
Update math.py

update example code of nanmean

Ligoml reviewed (approved):
LGTM for docs

XiaoguangHu01 reviewed (approved):
LGTM
